src/Server/ServerHelpers.ts

```diff
@@ -66,6 +66,199 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
+	 * First of, the names will be shortened for ease of manipulation:
+	 * n:= targetMoney (n for new), o:= startMoney (o for old), b:= exponentialBase, t:= threadMultiplier, c:= cycles/threads
+	 * c is what we are trying to compute.
+	 *
+	 * After growing, the money on a server is n = (o + c) * b^(c*t)
+	 * c appears in an exponent and outside it, this is usually solved using the productLog/lambert's W special function
+	 * this function will be noted W in the following
+	 * The idea behind lambert's W function is W(x)*exp(W(x)) = x, or in other words, solving for y, y*exp(y) = x, as a function of x
+	 * This function is provided in some advanced math library but we will compute it ourself here.
+	 *
+	 * Let's get back to solving the equation. It cannot be rewrote using W immediately because the base of the exponentiation is b
+	 * b^(c*t) = exp(ln(b)*c*t) (this is how a^b is defined on reals, it matches the definition on integers)
+	 * so n = (o + c) * exp(ln(b)*c*t) , W still cannot be used directly. We want to eliminate the other terms in 'o + c' and 'ln(b)*c*t'.
+	 *
+	 * A change of variable will do. The idea is to add an equation introducing a new variable (w here) in the form c = f(w) (for some f)
+	 * With this equation we will eliminate all references to c, then solve for w and plug the result in the new equation to get c.
+	 * The change of variable performed here should get rid of the unwanted terms mentionned above, c = w/(ln(b)*t) - o should help.
+	 * This change of variable is allowed because whatever the value of c is, there is a value of w such that this equation holds:
+	 * w = (c + o)*ln(b)*t  (see how we used the terms we wanted to eliminate in order to build this variable change)
+	 *
+	 * We get n = (o + w/(ln(b)*t) - o) * exp(ln(b)*(w/(ln(b)*t) - o)*t) [ = w/(ln(b)*t) * exp(w - ln(b)*o*t) ]
+	 * The change of variable exposed exp(w - o*ln(b)*t), we can rewrite that with exp(a - b) = exp(a)/exp(b) to isolate 'w*exp(w)'
+	 * n = w/(ln(b)*t) * exp(w)/exp(ln(b)*o*t) [ = w*exp(w) / (ln(b) * t * b^(o*t)) ]
+	 * Almost there, we just need to cancel the denominator on the right side of the equation:
+	 * n * ln(b) * t * b^(o*t) = w*exp(w), Thus w = W(n * ln(b) * t * b^(o*t))
+	 * Finally we invert the variable change: c = W(n * ln(b) * t * b^(o*t))/(ln(b)*t) - o
+	 *
+	 * There is still an issue left: b^(o*t) doesn't fit inside a double precision float
+	 * because the typical amount of money on servers is arround 10^6~10^9
+	 * We need to get an approximation of W without computing the power when o is huge
+	 * Thankfully an approximation giving ~30% error uses log immediately so we will use
+	 * W(n * ln(b) * t * b^(o*t)) ~= log(n * ln(b) * t * b^(o*t)) = log(n * ln(b) * t) + log(exp(ln(b) * o * t))
+	 * = log(n * ln(b) * t) + ln(b) * o * t
+	 * (thanks to Drak for the grow formula, f4113nb34st and Wolfram Alpha for the rewrite, dwRchyngqxs for the explanation)
```

Comment:
I have implemented [`WLog(x) = W(ln(x))`](https://github.com/howion/lambert-w-function/issues/6) which allows us to handle `b^(o*t)` with ease. So `W(n * ln(b) * t * b^(o*t))` becomes `WLog(ln(n * ln(b) * t) + o * t * ln(b))`.

Reply:
I will reuse [your code](https://gist.github.com/xmodar/baa392fc2bec447d10c2c20bbdcaf687) of the initial approximation given in [Iacono, R., Boyd, J.P.](https://doi.org/10.1007/s10444-017-9530-3) (20). I doubt I will use their iteration on lambert W though because the code is intended to be read by people and maybe teach them something, like using results from research articles and a classical algorithm (Newton's iteration).

Reply:
My very first solution to this problem was actually Newton's method. Then, I attempted using an arbitrary-precision floating-point library like [mpmath](https://mpmath.org/doc/current/functions/powers.html?highlight=lambert#mpmath.lambertw). However, I found [the paper](https://doi.org/10.1007/s10444-017-9530-3) and their solution seems to work well but suffers from the same issue. Therefore, I thought about the log scale and had to derive this implementation myself. The idea behind it is to leverage the [LogSumExp trick](https://en.wikipedia.org/wiki/LogSumExp#:~:text=log-domain%20calculations) to define `log1Exp(x) = ln(1 + exp(x))`. The rest is a bunch of careful substitutions to reach the final form.

P.S. I just update [the gist](https://gist.github.com/xmodar/baa392fc2bec447d10c2c20bbdcaf687) to improve handling boundary conditions.